src/fstext/fstext-utils-inl.h

```diff
@@ -114,6 +114,55 @@ void GetInputSymbols(const Fst<Arc> &fst,
   std::sort(symbols->begin(), symbols->end());
 }
 
+template<class Arc, class I>
+class RemoveSomeInputSymbolsMapper {
+public:
+  Arc operator ()(const Arc &arc_in) {
+    Arc ans = arc_in;
+    if (to_remove_set_.count(ans.ilabel) != 0) ans.ilabel = 0;  // remove this symbol
+    return ans;
+  }
+  MapFinalAction FinalAction() { return MAP_NO_SUPERFINAL; }
+  MapSymbolsAction InputSymbolsAction() { return MAP_CLEAR_SYMBOLS; }
+  MapSymbolsAction OutputSymbolsAction() { return MAP_COPY_SYMBOLS; }
+  uint64 Properties(uint64 props) const {
+    // remove the following as we don't know now if any of them are true.
+    uint64 to_remove = kAcceptor|kNotAcceptor|kIDeterministic|kNonIDeterministic|
+        kNoEpsilons|kNoIEpsilons|kILabelSorted|kNotILabelSorted;
+    return props & ~to_remove;
+  }
+  RemoveSomeInputSymbolsMapper(const std::vector<I> &to_remove):
+      to_remove_set_(to_remove) {
+    KALDI_ASSERT_IS_INTEGER_TYPE(I);
+         assert(to_remove_set_.count(0) == 0);  // makes no sense to remove epsilon.
+       }
+private:
+  kaldi::ConstIntegerSet<I> to_remove_set_;
+};
+
+template<class Arc, class I>
+using LookaheadFst = ArcMapFst<Arc, Arc, RemoveSomeInputSymbolsMapper<Arc, I> >;
+
+// Lookahead composition is used for optimized online
+// composition of FSTs during decoding. See
+// nnet3/nnet3-latgen-faster-lookahead.cc
+template<class Arc, class I>
+LookaheadFst<Arc, I> LookaheadComposeFst(const Fst<Arc> &ifst1,
+                                         const Fst<Arc> &ifst2,
+                                         const std::vector<I> &to_remove) {
+  typedef LookAheadMatcher< StdFst > M;
```

Comment:
How did you know that these were the right filters to use? Did you copy-paste from some sample inside OpenFST? I just want to double-check. I've read most of the paper, but I am not familiar with this part of the OpenFST codebase.

Reply:
That all from the sources. I actually just changed it to more high-level openfst API, same effect as before.